pandas/_libs/groupby_mode_helper.pxi.in

```diff
@@ -0,0 +1,177 @@
+    #khcomplex128_t,
+    khiter_t,
+)
+
+from pandas._libs.hashtable import (
+    # NaN checking
+    #is_nan_khcomplex128_t,
+    #is_nan_khcomplex64_t,
+    is_nan_float64_t,
+    is_nan_float32_t,
+    is_nan_int64_t,
+    is_nan_int32_t,
+    is_nan_int16_t,
+    is_nan_int8_t,
+    is_nan_uint64_t,
+    is_nan_uint32_t,
+    is_nan_uint16_t,
+    is_nan_uint8_t,
+    # Casting
+    #to_complex64,
+    #to_complex128,
+    #to_khcomplex128_t,
+    #to_khcomplex64_t,
+)
+
+{{py:
+# TODO: add complex64 and complex128 (requires comparisons between complex numbers)
+# dtype, ttype, c_type, to_c_type, to_dtype
+dtypes = [#('complex128', 'complex128', 'khcomplex128_t',
+          #               'to_khcomplex128_t', 'to_complex128'),
+          #('complex64', 'complex64', 'khcomplex64_t',
+          #              'to_khcomplex64_t', 'to_complex64'),
+          ('float64', 'float64', 'float64_t', '', ''),
+          ('float32', 'float32', 'float32_t', '', ''),
+          ('uint64', 'uint64', 'uint64_t', '', ''),
+          ('uint32', 'uint32', 'uint32_t', '', ''),
+          ('uint16', 'uint16', 'uint16_t', '', ''),
+          ('uint8', 'uint8', 'uint8_t', '', ''),
+          ('object', 'pymap', 'object', '', ''),
+          ('int64', 'int64', 'int64_t', '', ''),
+          ('int32', 'int32', 'int32_t', '', ''),
+          ('int16', 'int16', 'int16_t', '', ''),
+          ('int8', 'int8', 'int8_t', '', '')]
+
+}}
+
+{{for dtype, ttype, c_type, to_c_type, to_dtype in dtypes}}
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef {{c_type}} calc_mode_{{dtype}}(kh_{{ttype}}_t *table):
+    cdef:
+        {{c_type}} mode = 0 # fix annoying uninitialized warning
+        {{c_type}} val
+        int count, max_count = 0
+        khiter_t k
+
+    for k in range(table.n_buckets):
+        if kh_exist_{{ttype}}(table, k):
+            count = table.vals[k]
+            {{if dtype != 'object'}}
+            val = table.keys[k]
+            if count == max_count and val < mode:
+            {{else}}
+            val = <object>table.keys[k]
+            if count == max_count:
+            {{endif}}
+                mode = val
+            elif count > max_count:
+                mode = val
+                max_count = count
+    return mode
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def group_mode_{{dtype}}(ndarray[{{c_type}}, ndim=1] out,
+                         ndarray[{{c_type}}, ndim=1] values,
+                         ndarray[int64_t, ndim=1] labels,
+                         bint dropna = True):
+    """
+    Calculates the mode of each group.
+    If multimodal returns the smallest mode in each group if numeric.
+    For all other datatypes, returns a mode.
+    """
+    cdef:
+        Py_ssize_t i, N = len(values)
+        int64_t lab, curr_label = -1
+        kh_{{ttype}}_t *table
+        khiter_t k
+        int ret = 0
+
+    table = kh_init_{{ttype}}()
+    {{if dtype != 'object'}}
+    #TODO: Fix NOGIL later
```

Comment:
The non-numeric code paths prevent the Gil from being released. At least until Cython 3 gets released we usually do something like this:

https://github.com/pandas-dev/pandas/blob/859a78729c3b670136093b749912c2d752664203/pandas/_libs/groupby.pyx#L928